pallene/to_ir.lua

```diff
@@ -209,66 +209,150 @@ function ToIR:convert_stat(cmds, stat)
-        --   a_any, b_any = iter(st, ctrl)
-        --   if a_any == nil then break end
-        --   local a = a_any as T1
-        --   local b = b_any as T2
-        --   <loop body>
-        -- end
-        -- ```
-
-        local v_iter = ir.add_local(self.func, "$iter", exps[1]._type)
-        self:exp_to_assignment(cmds, v_iter, exps[1])
-        local v_state = ir.add_local(self.func, "$st", exps[2]._type)
-        self:exp_to_assignment(cmds, v_state, exps[2])
-
-        local v_lhs_dyn = {}
-        for _, decl in ipairs(decls) do
-            local v = ir.add_local(self.func, "$" .. decl.name .. "_dyn", types.T.Any())
-            table.insert(v_lhs_dyn, v)
-        end
+        local e1 = exps[1]
+        local is_ipairs = (
+            e1._tag == "ast.Exp.CallFunc" and
+            e1.exp._tag == "ast.Exp.Var" and
+            e1.exp.var._name._tag == "checker.Name.Builtin" and
+            e1.exp.var._name.name == "ipairs")
+
+
+        if is_ipairs then
+            -- `ipairs` are desugared down to regular for-loops
+            -- ```
+            -- for i: T1, x: T2 in ipairs(xs) do
+            --   <loop body>
+            -- end
+            -- ```
+            -- would get compiled down to:
+            -- ```
+            -- local i_num: integer = 1
+            -- while true do
+            --   local x_dyn = xs[i_num]
+            --   if x_dyn == nil then
+            --     break
+            --   end
+            --   local i = i_num as T1
+            --   local x = x_dyn as T2 
+            --   <loop body>
+            --   i_num = i_num + 1
+            -- end
+            -- ```
+
+            -- the table passed as argument to `ipairs`
+            local arr =  exps[2].call_exp.args[1]
+            assert(arr._type.elem._tag == "types.T.Any")
+            local v_arr = ir.add_local(self.func, "$xs", arr._type)
+            self:exp_to_assignment(cmds, v_arr, arr)
+
+            -- local i_num: integer = 1
+            local v_inum = ir.add_local(self.func, "$"..decls[1].name.."_num", types.T.Integer())
+            local start = ir.Value.Integer(1)
+            table.insert(cmds, ir.Cmd.Move(stat.loc, v_inum, start))
+        
+            -- body of the while loop.
+            local body = {}
+
+            -- x_dyn = xs[i_num]
+            local v_x_dyn = ir.add_local(self.func, "$"..decls[2].name.."_dyn", types.T.Any())
+            local src_arr =  ir.Value.LocalVar(v_arr)
+            local src_i =  ir.Value.LocalVar(v_inum)
+            table.insert(body, ir.Cmd.GetArr(stat.loc, types.T.Any(), v_x_dyn, src_arr, src_i))
+
+            -- if x_dyn == nil then break end
+            local v_cond_checknil = ir.add_local(self.func, "$is_"..decls[2].name.."_nil", types.T.Boolean())
+            table.insert(body, ir.Cmd.IsNil(stat.loc, v_cond_checknil, ir.Value.LocalVar(v_x_dyn)))
+            table.insert(body, ir.Cmd.If(stat.loc, ir.Value.LocalVar(v_cond_checknil), ir.Cmd.Break(), ir.Cmd.Nop()))
```

Comment:
Since the `_is_nil` variable is only used in the next command after this one, I think it's OK if we just don't give a name for it. That is, pass `false` as the second argument to `add_local`. This should also avoid the issue with `decls[2]` that I previously mentioned.